File: app/src/main/java/com/kksionek/gdzietentramwaj/view/FavoriteLinesActivity.java
package com.kksionek.gdzietentramwaj.view;

import android.arch.lifecycle.ViewModelProviders;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.GridView;
import android.widget.TextView;

import com.kksionek.gdzietentramwaj.DataSource.Room.FavoriteTram;
import com.kksionek.gdzietentramwaj.R;
import com.kksionek.gdzietentramwaj.ViewModel.FavoriteLinesActivityViewModel;

import java.util.Collections;
import java.util.List;

import io.reactivex.Observable;
import io.reactivex.schedulers.Schedulers;

public class FavoriteLinesActivity extends AppCompatActivity {

    private FavoriteLinesActivityViewModel mViewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_favorite_lines);

        Toolbar myToolbar = findViewById(R.id.my_toolbar);
        setSupportActionBar(myToolbar);
        getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        getSupportActionBar().setDisplayShowHomeEnabled(true);
        myToolbar.setNavigationOnClickListener(v -> onBackPressed());

        GridView gridView = findViewById(R.id.gridView);

        mViewModel = ViewModelProviders.of(this)
                .get(FavoriteLinesActivityViewModel.class);

        FavoritesAdapter adapter = new FavoritesAdapter();
        gridView.setAdapter(adapter);
        mViewModel.getFavoriteTrams().observe(this, adapter::setData);
    }

    private class FavoritesAdapter extends ArrayAdapter<FavoriteTram> {

        FavoritesAdapter() {
            super(FavoriteLinesActivity.this, R.layout.grid_favorite_element);
        }

        @Override
        @NonNull
        public View getView(int position, View convertView, @NonNull ViewGroup parent) {
            ViewHolder holder;
            if (convertView == null) {
                convertView = LayoutInflater.from(FavoriteLinesActivity.this).inflate(
                        R.layout.grid_favorite_element,
                        parent,
                        false);

                holder = new ViewHolder();
                holder.textView = convertView.findViewById(R.id.tramNum);
                convertView.setTag(holder);
            } else {
                holder = (ViewHolder) convertView.getTag();
            }

            final FavoriteTram tramData = getItem(position);
            holder.textView.setText(tramData.getLineId());
            if (tramData.isFavorite()) {
                convertView.setBackgroundResource(R.color.favoriteLineColor);
            } else {
                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                    convertView.setBackground(null);
                } else {
                    convertView.setBackgroundDrawable(null);
                }
            }

            convertView.setOnClickListener(
                    v -> Observable.fromCallable(() -> {
                        mViewModel.setTramFavorite(tramData.getLineId(), !tramData.isFavorite());
                        return 1;
                    }).subscribeOn(Schedulers.io())
                    .subscribe());
            return convertView;
        }

        public void setData(List<FavoriteTram> favoriteTrams) {
            clear();
            Collections.sort(favoriteTrams, new NaturalOrderComparator<>());
            addAll(favoriteTrams);
        }

        private class ViewHolder {
            TextView textView;
        }
    }
}

File: app/src/main/java/com/kksionek/gdzietentramwaj/favorite/view/FavoriteLinesActivity.kt
package com.kksionek.gdzietentramwaj.favorite.view

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.GridLayoutManager
import android.view.MenuItem
import com.kksionek.gdzietentramwaj.R
import com.kksionek.gdzietentramwaj.TramApplication
import com.kksionek.gdzietentramwaj.base.dataSource.FavoriteTram
import com.kksionek.gdzietentramwaj.base.viewModel.ViewModelFactory
import com.kksionek.gdzietentramwaj.favorite.viewModel.FavoriteLinesActivityViewModel
import kotlinx.android.synthetic.main.activity_favorite_lines.*
import javax.inject.Inject


private const val COLUMN_COUNT = 7

class FavoriteLinesActivity : AppCompatActivity() {

    // https://medium.com/androiddevelopers/testing-room-migrations-be93cdb0d975

    @Inject
    internal lateinit var viewModelFactory: ViewModelFactory

    private lateinit var mViewModel: FavoriteLinesActivityViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_favorite_lines)

        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        supportActionBar?.setHomeButtonEnabled(true)

        (application as TramApplication).appComponent.inject(this)

        mViewModel = ViewModelProviders.of(this, viewModelFactory)
            .get(FavoriteLinesActivityViewModel::class.java)

        val adapter = FavoritesAdapter {
            mViewModel.setTramFavorite(it.lineId, !it.isFavorite)
        }
        gridView.layoutManager = GridLayoutManager(this, COLUMN_COUNT)
        gridView.addItemDecoration(SpacesItemDecoration(this, R.dimen.grid_offset))
        gridView.itemAnimator = object : DefaultItemAnimator() {
            override fun getChangeDuration(): Long = 100
        }
        gridView.adapter = adapter
        mViewModel.favoriteTrams
            .observe(this, Observer<List<FavoriteTram>> { adapter.submitList(it) })
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }
}
